test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java

```diff
@@ -0,0 +1,112 @@
+ * @run main/othervm NegativeArraySizeTest
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InvalidClassException;
+import java.io.IOException;
+import java.io.ObjectInputFilter;
+import java.io.ObjectInputFilter.Status;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamException;
+
+public class NegativeArraySizeTest {
+
+    private static byte[] _buildPayload() throws IOException {
+        String[] simpleArray = new String[1];
+
+        // Serialize to bytes
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
+        oos.writeObject(simpleArray);
+        oos.close();
+        byte[] serializedData = baos.toByteArray();
+
+        // Find the right location to modify, looking for the first instance of TC_ENDBLOCKDATA
+        int firstPos = 0;
+        for (int i=0; i<serializedData.length-1; i++) {
+            // 0x78 = TC_ENDBLOCKDATA
+            if (serializedData[i] == 0x78) {
+                firstPos = i;
+                break;
+            }
+        }
+
+        // Replace array length with -2
+        serializedData[firstPos+2] = (byte) 0xff;
+        serializedData[firstPos+3] = (byte) 0xff;
+        serializedData[firstPos+4] = (byte) 0xff;
+        serializedData[firstPos+5] = (byte) 0xfe;
+
+        return  serializedData;
+    }
+
+    public static void test_simpleArray_negative() throws IOException, ClassNotFoundException {
+        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(_buildPayload()));
+        ois.readObject();
+    }
+
+    public static void test_simpleArray_negative_with_filter() throws IOException, ClassNotFoundException {
+        class CustomFilter implements ObjectInputFilter {
+            @Override
+            public Status checkInput(FilterInfo filterInfo) {
+                Class<?> cl = filterInfo.serialClass();
+                if (cl != null && cl.isArray() && filterInfo.arrayLength() < -1) {
+                    throw new RuntimeException("FilterInfo.arrayLength() must be >= -1 for arrays (was " + filterInfo.arrayLength() + ")");
+                }
+                return Status.ALLOWED;
+            }
+        }
+        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(_buildPayload()));
+        ois.setObjectInputFilter(new CustomFilter());
+        ois.readObject();
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
+            test_simpleArray_negative();
+        } catch (NegativeArraySizeException nase) {
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException
```

Comment:
Check the message from `ose` here?

Reply:
Done. I only hope nobody will localize the exception message :)